File: src/chipschallenge/buttonbehaviors/ToggleWallBehavior.java
package chipschallenge.buttonbehaviors;

import chipschallenge.Block;
import chipschallenge.Block.Type;
import chipschallenge.Game;

/**
 * If the button was green and the listener is a toggle wall. Toggle it.
 */
public class ToggleWallBehavior implements ButtonBehavior {

    private ToggleWallBehavior() {
    }
    private static ToggleWallBehavior mInstance = null;

    public static synchronized ToggleWallBehavior getInstance() {
        if (mInstance == null) {
            mInstance = new ToggleWallBehavior();
        }
        return mInstance;
    }

    public void buttonDown(Block listener, Block button) {
        Game.getInstance().moveOccured(listener.getPoint());
        if (listener.isA(Block.Type.TOGGLEWALLOPEN)) {
            listener.replace(Block.create(Type.TOGGLEWALLCLOSED));
        } else if (listener.isA(Type.TOGGLEWALLCLOSED)) {
            listener.replace(Block.create(Type.TOGGLEWALLOPEN));
        }
    }

    public void buttonUp(Block listener, Block button) {
        // Do nothing
    }
}

File: ChipsChallenge.Shared/Buttonbehaviors/ToggleWallBehavior.cs
namespace ChipsChallenge.Shared.Buttonbehaviors
{
    /// <summary>
    /// If the button was green and the listener is a toggle wall. Toggle it.
    /// </summary>
    public class ToggleWallBehavior : IButtonBehavior
    {
        private static ToggleWallBehavior instance;

        private ToggleWallBehavior()
        {
        }

        public static ToggleWallBehavior Instance
        {
            get
            {
                lock (typeof(ToggleWallBehavior))
                {
                    return instance ?? (instance = new ToggleWallBehavior());
                }
            }
        }

        public virtual void ButtonDown(Block listener, Block button)
        {
            Game.Instance.MoveOccured(listener.Point);
            if (listener.IsA(Block.Type.TOGGLEWALLOPEN))
            {
                listener.Replace(Block.Create(Block.Type.TOGGLEWALLCLOSED));
            }
            else if (listener.IsA(Block.Type.TOGGLEWALLCLOSED))
            {
                listener.Replace(Block.Create(Block.Type.TOGGLEWALLOPEN));
            }
        }

        public virtual void ButtonUp(Block listener, Block button)
        {
        }
    }
}